### PR TITLE
Add Hibernate 6 module

### LINKED ISSUE #158: Module for Hibernate 6
Are there any plans for development of a module for Hibernate 6 or rather 6.1 yet?

### PR DESCRIPTION
**Pull request by reda-alaoui**
Fix #158

Although it seemed possible to use `jackson-datatype-hibernate5-jakarta:2.14.1` with Hibernate 6, I had one failure with one of my tests:
```java
java.lang.NoSuchMethodError: 'java.io.Serializable org.hibernate.proxy.LazyInitializer.getIdentifier()'

	at com.fasterxml.jackson.datatype.hibernate5.jakarta.Hibernate5JProxySerializer.findProxied(Hibernate5JProxySerializer.java:219)
	at com.fasterxml.jackson.datatype.hibernate5.jakarta.Hibernate5JProxySerializer.serialize(Hibernate5JProxySerializer.java:137)
	at com.fasterxml.jackson.datatype.hibernate5.jakarta.Hibernate5JProxySerializer.serialize(Hibernate5JProxySerializer.java:39)
	at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:733)
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:774)
	at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:178)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:480)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.
```

At first, I thought excluding `org.hibernate:hibernate-core-jakarta` and `org.hibernate:hibernate-entitymanager` would be enough but it was not.
The only way to fix this for me was to use the module introduced by this PR.

This new hibernate6 module is a copy paste of `hibernate5-jakarta`  with the correct dependencies.

### CODE REVIEW

**Review by cowtowncoder — APPROVED** [review]

**Comment by cowtowncoder** [comment]
Sounds good @reda-alaoui -- thank you for providing this contribution!

Just one request: could you also modify `README.md` with a note of this being added for 2.15? I can merge this then.

**Comment by cowtowncoder** [comment]
@reda-alaoui Actually one more question/thought: looks like this is "Jakarta" variant of things. This makes sense.
But given that with Hibernate 5 we have "regular" and "jakarta" versions, should we follow that naming here, and make it `hibernate6-jakarta` (artifact) and `Hibernate6JakartaModule.java` (module class)?
We could keep package name as `com.fasterxml.jackson.datatype.hibernate6` without `jakarta` in there however; ditto for classnames (unlike Hibernate 5 which has `J` as in `Hibernate5JAnnotationIntrospector`)

WDYT?

**Comment by reda-alaoui** [comment]
Hello @cowtowncoder ,

> But given that with Hibernate 5 we have "regular" and "jakarta" versions, should we follow that naming here, and make it hibernate6-jakarta (artifact) and Hibernate6JakartaModule.java (module class)?

Hibernate 6 is only compatible with `jakarta.persistence`. Unlike Hibernate 5, It can't work with `javax.persistence`. 
So IMO, the new module name should not be suffixed with `-jarkata`.

**Comment by reda-alaoui** [comment]
> Sounds good @reda-alaoui -- thank you for providing this contribution!
> 
> Just one request: could you also modify `README.md` with a note of this being added for 2.15? I can merge this then.

README updated :+1:

File: hibernate6/src/moditect/module-info.java
Comment (cowtowncoder) [inline hibernate6/src/moditect/module-info.java:14]: Should be `Hibernate6Module`.

File: hibernate6/src/test/java/com/fasterxml/jackson/datatype/hibernate6/BaseTest.java
Comment (cowtowncoder) [inline hibernate6/src/test/java/com/fasterxml/jackson/datatype/hibernate6/BaseTest.java:13]: Let's remove these.

File: hibernate6/src/test/java/com/fasterxml/jackson/datatype/hibernate6/BaseTest.java
Comment (cowtowncoder) [inline hibernate6/src/test/java/com/fasterxml/jackson/datatype/hibernate6/BaseTest.java:25]: Let's convert to using builder:

    JsonMapper.builder().addModule(...).build();

in case there will be Jackson 3.x version of module

**Review by cowtowncoder — CHANGES_REQUESTED** [review]
Ok looks mostly good, couple of things (important one: module name)

File: hibernate6/pom.xml
Comment (cowtowncoder) [inline hibernate6/pom.xml:96]: I don't think this is needed as we don't include 5.2 test: let's remove `maven-jar-plugin` from here.

**Review by cowtowncoder — CHANGES_REQUESTED** [review]
Ok looks good, only one more thing to remove -- my bad, should have noticed earlier.
I think it was just copied from `hibernate5-jakarta` and I should remove it from there too.
Thanks!

**Comment by reda-alaoui** [comment]
Ping :)

**Comment by pjfanning** [comment]
If this change involves copy/pasting lots of code and compiling it with different dependencies then I would argue that we should use some build scripting to achieve this instead of maintaining 2 independent copies of the same code.

**Comment by cowtowncoder** [comment]
@pjfanning If anyone wants to do that, I am +1. Until then I am fine in copy-pasting since the module is without active maintainer and is not currently planned to be ported for Jackson 3.0. That is: I won't hold this PR for improvement work that would be desirable but big undertaking not directly related to problem at that.

**Review by cowtowncoder — APPROVED** [review]

**Comment by cowtowncoder** [comment]
Ok. Fudge. JDK baseline for building this module increases to JDK 11 it seems. Need to note on README

**Comment by cowtowncoder** [comment]
@reda-alaoui Build `mvn clean install` actually fails with JDK 11. How did you manage to get it built without failures?

**Comment by reda-alaoui** [comment]
@cowtowncoder If I remember correctly, Hibernate 5 tests were already failing when I started to work on this PR.

**Comment by cowtowncoder** [comment]
I get failures for Hibernate 4. 

I may have to revert this addition if I cannot build other modules for release...

**Comment by cowtowncoder** [comment]
Ok, JDK 11 is triggering the issue, unfortunately. It does indeed fail Hibernate 5 tests.
But Hibernate 6 must be built (and used) with JDK 11. So we have a problem.

Now... if there was a way to make Hib5/Hib4 tests pass on JDK 11 I would be ok building with it (targetting Java 8).
But I cannot break existing versions to get Hib6 to be included.
Unfortunately I have neither time nor expertise to really figure out failures -- but if anyone else has and can, will try my best to help with process things.

**Pull request merged**
